src/main/java/at/ac/tuwien/kr/alpha/common/depgraph/DependencyGraph.java

```diff
@@ -0,0 +1,202 @@
+	}
+
+	private Node handleRuleHead(NonGroundRule rule) {
+		LOGGER.trace("Processing head of rule: {}", rule);
+		Node retVal;
+		Predicate pred;
+		if (rule.isConstraint()) {
+			pred = this.generateConstraintDummyPredicate();
+			retVal = new Node(pred, pred.toString(), true);
+			List<Edge> dependencies = new ArrayList<>();
+			dependencies.add(new Edge(retVal, false, "-"));
+			if (this.nodes.containsKey(retVal)) {
+				throw new IllegalStateException("Dependency graph already contains node for constraint " + pred.toString() + "!");
+			}
+			this.nodes.put(retVal, dependencies);
+			this.nodesByPredicate.put(pred, retVal);
+		} else {
+			Atom head = rule.getHeadAtom();
+			pred = head.getPredicate();
+			retVal = new Node(pred, pred.toString());
+			if (!this.nodesByPredicate.containsKey(pred)) {
+				this.nodes.put(retVal, new ArrayList<>());
+				this.nodesByPredicate.put(pred, retVal);
+			} else {
+				retVal = this.getNodeForPredicate(pred);
+			}
+		}
+		return retVal;
+	}
+
+	private void handleRuleBodyLiteral(Node headNode, Literal lit) {
+		List<Edge> dependants;
+		Edge tmpEdge;
+		Node bodyNode;
+		Predicate p = lit.getPredicate();
+		if (!this.nodesByPredicate.containsKey(p)) {
+			LOGGER.trace("Creating new node for predicate {}", p);
+			dependants = new ArrayList<>();
+			bodyNode = new Node(p);
+			this.nodesByPredicate.put(p, bodyNode);
+			this.nodes.put(bodyNode, dependants);
+		} else {
+			LOGGER.trace("Node for predicate {} already exists", p);
+			bodyNode = this.getNodeForPredicate(p);
+			dependants = this.nodes.get(bodyNode);
+		}
+		tmpEdge = new Edge(headNode, !lit.isNegated(), lit.isNegated() ? "-" : "+");
+		if (!dependants.contains(tmpEdge)) {
+			LOGGER.trace("Adding dependency: {} -> {} ({})", bodyNode.getPredicate(), headNode.getPredicate(), tmpEdge.getSign() ? "+" : "-");
+			dependants.add(tmpEdge);
+		}
+		this.nodesByPredicate.put(lit.getPredicate(), bodyNode);
+	}
+
+	/**
+	 * (Re-)Writes the <code>transposedNodes</code> of this <code>DependencyGraph</code>
+	 */
+	private void buildTransposedStructure() {
+		Map<Node, List<Edge>> transposed = new HashMap<>();
+		Node srcNode;
+		Node targetNode;
+		for (Map.Entry<Node, List<Edge>> entry : this.nodes.entrySet()) {
+			srcNode = entry.getKey();
+			if (!transposed.containsKey(srcNode)) {
+				transposed.put(srcNode, new ArrayList<>());
+			}
+			for (Edge e : entry.getValue()) {
+				targetNode = e.getTarget();
+				if (!transposed.containsKey(e.getTarget())) {
+					transposed.put(targetNode, new ArrayList<>());
+				}
+				transposed.get(targetNode).add(new Edge(entry.getKey(), e.getSign(), e.getLabel()));
+			}
+		}
+		this.transposedNodes = transposed;
+	}
+
+	private Predicate generateConstraintDummyPredicate() {
+		Predicate retVal = Predicate.getInstance(String.format(DependencyGraph.CONSTRAINT_PREDICATE_FORMAT, this.nextConstraintNumber()), 0);
+		return retVal;
+	}
+
+	// TODO maybe move this to DependencyGraphUtils
```

Comment:
Please address this todo. Since this method relies a lot on `DependencyGraphUtils` I guess it could be part of that class too.